File: area_reg.m
function [res,dSdx,dSdy,dSdz]=area_reg(tlist,vlist)
%Regularization function
%res(j)=(A_j-1/n)/A
nfac=size(tlist,1);
nvert=size(vlist,1);
areasum=0;
area=zeros(1,nfac);
areadx=zeros(nfac,nvert);
aready=zeros(nfac,nvert);
areadz=zeros(nfac,nvert);
for j=1:nfac
    v1=vlist(tlist(j,2),:)-vlist(tlist(j,1),:);
    v2=vlist(tlist(j,3),:)-vlist(tlist(j,1),:);
    x1=vlist(tlist(j,1),1);
    y1=vlist(tlist(j,1),2);
    z1=vlist(tlist(j,1),3);
    x2=vlist(tlist(j,2),1);
    y2=vlist(tlist(j,2),2);
    z2=vlist(tlist(j,2),3);
    x3=vlist(tlist(j,3),1);
    y3=vlist(tlist(j,3),2);
    z3=vlist(tlist(j,3),3);
    cp=cross(v1,v2);
    normc=norm(cp);
      area(j)=1/2*normc;
    
    cpdx=[0,z3-z2,y2-y3; 0,z1-z3,y3-y1;0,z2-z1,y1-y2];
    cpdy=[z2-z3,0,x3-x2; z3-z1,0,x1-x3;z1-z2,0,x2-x1];
    cpdz=[y3-y2,x2-x3,0;y1-y3,x3-x1,0;y2-y1,x1-x2,0];
    normcdx=zeros(1,3);
    normcdy=zeros(1,3);
    normcdz=zeros(1,3);
%     for k=1:3
%         normcdx(k)=1/normc*dot(cp,cpdx(k,:));
%         normcdy(k)=1/normc*dot(cp,cpdy(k,:));
%         normcdz(k)=1/normc*dot(cp,cpdz(k,:));
%     end
    normcdx=1/normc*cp*cpdx';
normcdy=1/normc*cp*cpdy';
normcdz=1/normc*cp*cpdz';
    
    i1=tlist(j,1);
    i2=tlist(j,2);
    i3=tlist(j,3);
    areadx(j,[i1 i2 i3])=0.5*normcdx;
    aready(j,[i1 i2 i3])=0.5*normcdy;
    areadz(j,[i1 i2 i3])=0.5*normcdz;
    
    
%Sum of all facet areas

end
dSdx=zeros(nfac,nvert);
dSdy=zeros(nfac,nvert);
dSdz=zeros(nfac,nvert);
dAdx=sum(areadx);
dAdy=sum(aready);
dAdz=sum(areadz);
Asum=sum(area);
for j=1:nfac
    res(j)=(area(j)-1/nfac)/Asum;
    dSdx(j,:)=(areadx(j,:)*Asum-(area(j)-1/nfac)*dAdx)/Asum^2;
    dSdy(j,:)=(aready(j,:)*Asum-(area(j)-1/nfac)*dAdy)/Asum^2;
    dSdz(j,:)=(areadz(j,:)*Asum-(area(j)-1/nfac)*dAdz)/Asum^2;
    %dSdz(j,:)=areadz(j,:)-dAdz;
end
end

File: area_reg.c
#include"utils.h"
#include<stdlib.h>
#include<math.h>
#include"matrix_ops.h"
void area_reg(int *tlist,double *vlist,int nfac,int nvert,double *D,int dm,int dn,double *area,double *dArdv)
{
    /*Calculate area regularization, penalize divergence from mean area. area(j)=A_j/Sum(A)-1/n*/
    /*area: nfac array
     * dArdv: nfac x 3*nvert+3 array */
    /*NB ALLOCATE MEMORY BEFOREHAND*/
    double *v1,*v2,*v3,n[3];
    double dndx1[3],dndx2[3],dndx3[3];
    double dndy1[3],dndy2[3],dndy3[3];
    double dndz1[3],dndz2[3],dndz3[3];
    double *areadx,*aready,*areadz,*dArdx,*dArdy,*dArdz;
    areadx=calloc(nvert*nfac,sizeof(double));
    aready=calloc(nvert*nfac,sizeof(double));
    areadz=calloc(nvert*nfac,sizeof(double));
    double A;
    double dAdx[3],dAdy[3],dAdz[3];
    double TA=0;
    int j1,j2,j3;
    if(D!=NULL && dm!=nvert)
    {
        puts("Error: Number of vertex coordinates is not equal to the number of rows in D.");
        exit(1);
    }
    if(D==NULL)
        dn=nvert;
    for(int j=0;j<nfac;j++)
    {
        j1=tlist[3*j]-1; /*NB: WE INDEX FROM ZERO*/
        j2=tlist[3*j+1]-1;
        j3=tlist[3*j+2]-1;
        v1=vlist+3*j1;
        v2=vlist+3*j2;
        v3=vlist+3*j3;
        Calculate_Area_and_Normal_Derivative(v1,v2,v3,n,dndx1,dndx2,dndx3,dndy1,dndy2,dndy3,dndz1,dndz2,dndz3,&A,dAdx,dAdy,dAdz);
        TA=TA+A;
        area[j]=A;
        areadx[j*nvert+j1]=dAdx[0];
        areadx[j*nvert+j2]=dAdx[1];
        areadx[j*nvert+j3]=dAdx[2];
        
        aready[j*nvert+j1]=dAdy[0];
        aready[j*nvert+j2]=dAdy[1];
        aready[j*nvert+j3]=dAdy[2];
        
        areadz[j*nvert+j1]=dAdz[0];
        areadz[j*nvert+j2]=dAdz[1];
        areadz[j*nvert+j3]=dAdz[2];
    }
    //Divide with total area and calculate final derivatives
    //Derivative of total area wrt vertices
    double *dTAdx,*dTAdy,*dTAdz;
    dTAdx=malloc(nvert*sizeof(double));
    dTAdy=malloc(nvert*sizeof(double));
    dTAdz=malloc(nvert*sizeof(double));
    for(int k=0;k<nvert;k++)
    {
        dTAdx[k]=0;
        dTAdy[k]=0;
        dTAdz[k]=0;
        for(int j=0;j<nfac;j++)
        {
          dTAdx[k]+=areadx[j*nvert+k];
          dTAdy[k]+=aready[j*nvert+k];  
          dTAdz[k]+=areadz[j*nvert+k];
        }
    }
    if(D==NULL)
    {
        
    for(int j=0;j<nfac;j++)
    {
        
        for(int k=0;k<nvert;k++)
        {
            dArdv[j*(3*nvert+3)+k]=(areadx[j*nvert+k]*TA-area[j]*dTAdx[k])/pow(TA,2);
            dArdv[j*(3*nvert+3)+k+nvert]=(aready[j*nvert+k]*TA-area[j]*dTAdy[k])/pow(TA,2);
            dArdv[j*(3*nvert+3)+k+2*nvert]=(areadz[j*nvert+k]*TA-area[j]*dTAdz[k])/pow(TA,2);
        }
        area[j]=area[j]/TA-1.0/nfac;
    }
    
    }
    else
    {
        dArdx=calloc(nvert*nfac,sizeof(double));
        dArdy=calloc(nvert*nfac,sizeof(double));
        dArdz=calloc(nvert*nfac,sizeof(double));
        for(int j=0;j<nfac;j++)
    {
        
        for(int k=0;k<nvert;k++)
        {
            areadx[j*nvert+k]=(areadx[j*nvert+k]*TA-area[j]*dTAdx[k])/pow(TA,2);
            aready[j*nvert+k]=(aready[j*nvert+k]*TA-area[j]*dTAdy[k])/pow(TA,2);
            areadz[j*nvert+k]=(areadz[j*nvert+k]*TA-area[j]*dTAdz[k])/pow(TA,2);
        }
        area[j]=area[j]/TA-1.0/nfac;
    }
    matrix_prod(areadx,nfac,nvert,D,dn,dArdx);
    matrix_prod(aready,nfac,nvert,D,dn,dArdy);
    matrix_prod(areadz,nfac,nvert,D,dn,dArdz);
    
    set_submatrix(dArdv,nfac,3*dn+3,dArdx,nfac,dn,0,0);
    set_submatrix(dArdv,nfac,3*dn+3,dArdy,nfac,dn,0,dn);
    set_submatrix(dArdv,nfac,3*dn+3,dArdz,nfac,dn,0,2*dn);
     free(dArdx);
    free(dArdy);
    free(dArdz);
    }
   
    free(areadx);
    free(aready);
    free(areadz);
    free(dTAdx);
    free(dTAdy);
    free(dTAdz);
}
/* 
 int main()
 {
      int nfac=8,nfacn;
      int nvert=6,nvertn;
     int *tlist;
     double *vlist;
     double *vlistn;
     int *tlistn;
     double *D;
     char shapefile[]="shape2.txt";
     read_shape(shapefile,&tlist,&vlist,&nfac,&nvert,0);
      double ini_dims[]={90,90,70};
      mul_cols(vlist,nvert,3,ini_dims);
     Sqrt3_Subdiv(tlist,vlist,nfac,nvert,&tlistn,&vlistn,&nfacn,&nvertn,&D);
     int nedgen=nfacn+nvertn-2;
     int nedge=nfac+nvert-2;
     double res,*dresdv;
   
   
     double *area,*areadv;
     area=calloc(nfacn,sizeof(double));
     areadv=calloc(nfacn*(3*nvert+3),sizeof(double));
     
     area_reg(tlistn,vlistn,nfacn,nvertn,D,nvertn,nvert,area,areadv);
     //print_matrix(area,1,nfac);
    // print_matrix(areadv,nfac,3*nvert+3);
     
    // free(D);
     write_matrix_file("/tmp/area2.txt",area,1,nfacn);
    write_matrix_file("/tmp/areadv2.txt",areadv,nfacn,3*nvert+3);
    free(area);
     free(areadv);
 }
 
*/
